### DISCUSSION TITLE
How to prevent cmp from auto-selecting first entry on <enter>

### DISCUSSION THREAD

**Discussion by srinathava** [hrsh7th/nvim-cmp#997] [Q&A] [answered]
Hi,

I'm trying to prevent nvim-cmp from auto-inserting the first item shown in the completion menu when I press `<enter>`. Instead I would like to press `<C-n>` or some other key to force a selection and only then <enter> should select it.

I have nvim-cmp setup like this currently:

```vim
" In ~/.config/nvim/init.vim
set completeopt=menuone,noselect,noinsert
```

```lua
-- Setup nvim-cmp.
local cmp = require'cmp'

cmp.setup({
    snippet = {
        -- REQUIRED - you must specify a snippet engine
        expand = function(args)
            require('luasnip').lsp_expand(args.body) -- For `luasnip` users.
        end,
    },
    window = {
        completion = cmp.config.window.bordered(),
        documentation = cmp.config.window.bordered(),
    },
    mapping = cmp.mapping.preset.insert({
        ['<C-b>'] = cmp.mapping.scroll_docs(-4),
        ['<C-f>'] = cmp.mapping.scroll_docs(4),
        ['<C-Space>'] = cmp.mapping.complete(),
        ['<C-e>'] = cmp.mapping.abort(),
        ['<CR>'] = cmp.mapping.confirm({ select = false }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
    }),
    sources = cmp.config.sources({
        { name = 'nvim_lsp' },
        { name = 'luasnip' }, -- For luasnip users.
        { name = 'buffer' },
    }),
    completion = {
        completeopt = 'menu,menuone,noinsert'
    }
})
```

Please note the part about 
```
['<CR>'] = cmp.mapping.confirm({ select = false }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
```

However, with this setting, if I open a Python file with the contents:

```python
def foobar():
    y = foo
```

and go the end of the second line `y = foo` and type just `b<CR>`, it auto-selects and then auto-inserts `foobar`. Basically I have no way of just writing `y = foob` on the second line. I have to basically backspace over all the stuff which just got inserted.

I'd greatly appreciate any help fixing this :)

Alternatively, is there a good way to configure `cmp` so I can press something other than <enter> and it just inserts a literal "<CR>"?

**Comment by Shougo** [comment]
Please escape your text.  Some characters are hidden.

And `select = false` works for me.  Please describe it more detailed.  Why it does not work.

**Comment by Shougo** [comment, marked as answer]
Please see my minimal init.vim

```vim
set rtp+=~/src/nvim-cmp
set rtp+=~/src/cmp-buffer
set rtp+=~/src/cmp-nvim-lsp
set rtp+=~/src/vim-vsnip
set rtp+=~/src/cmp-vsnip
set rtp+=~/src/nvim-lspconfig
set rtp+=~/src/cmp-cmdline

set completeopt=menuone,noselect

lua <<EOF
  -- Setup nvim-cmp.
  local cmp = require'cmp'

  cmp.setup({
    snippet = {
      expand = function(args)
        vim.fn["vsnip#anonymous"](args.body)
      end,
    },
    mapping = {
      ['<CR>'] = cmp.mapping.confirm({ select = false }),
      ['<tab>'] = cmp.mapping(cmp.mapping.confirm({ select = true }), { 'i', 's', 'c' }),
      ['<C-n>'] = cmp.mapping({
          i = function(fallback)
              if cmp.visible() then
                  cmp.select_next_item({ behavior = cmp.SelectBehavior.Select })
              else
                  fallback()
              end
          end
      }),
      ['<C-p>'] = cmp.mapping({
          i = function(fallback)
              if cmp.visible() then
                  cmp.select_prev_item({ behavior = cmp.SelectBehavior.Select })
              else
                  fallback()
              end
          end
      }),
    },
    sources = {
      { name = 'nvim_lsp' },
      { name = 'buffer' },
    },
  })
  cmp.setup.cmdline(':', {
    mapping = cmp.mapping.preset.cmdline(),
    sources = {
      { name = 'cmdline' }
    }
  })
EOF

lua << EOF
local capabilities = require('cmp_nvim_lsp').update_capabilities(vim.lsp.protocol.make_client_capabilities())

--require'lspconfig'.sumneko_lua.setup{}
require'lspconfig'.cssls.setup{}
require'lspconfig'.pylsp.setup{}
EOF

runtime! after/plugin/*.lua
```

**Reply by srinathava** [reply]
Thank you so much for the minimal config. This helped me fix the issue I was seeing. It turns out that the behavior I was seeing was because of the following in my original `cmp.setup`:

```lua
    completion = {
        completeopt = 'menu,menuone,noinsert'
    }
```
With that setting, this is the behavior I saw. When I open a Python file with:

```python
def foobaz():
    pass

def foobar():
    y = foo
```

and I go the end of the line `y = foo` and in insert mode, press `b`, it opens the completion menu, but it has already selected `foobar`:

![image](https://user-images.githubusercontent.com/61844/168942387-04f9d4f0-bd39-4123-9b72-9c7c54017ca3.png)

Then upon pressing `<CR>`, it immediately inserts `foobar`. Without that section in `cmp.setup`, I now get the correct behavior, pressing `b` only shows the items, but does not select `foobar`:

![image](https://user-images.githubusercontent.com/61844/168942659-c0efd714-a626-4caf-bc25-2cebb9519dfc.png)

Now pressing `<CR>` just inserts a literal `<CR>` into the text, which is exactly what I want! 

PS: Thanks for pointing out the hidden characters. I edited the question to fix them.

**Reply by Shougo** [reply]
Ah, OK.  `noinsert` means `select the first entry`.
So the `CR` inserts the first entry.
You need to use `noselect` instead.
